Allow click+drag to rearrange skills

zqsz-xiv opened the pull request:
Resolves #201. Also resolves #183 because I got annoyed while testing.

Clicking on the background of the timeline canvas, then dragging the cursor will create a selection box to select multiple skills (currently only achievable with shift + click on exact skill icons).

Clicking and dragging skills in the timeline editor and canvas now will preview the drop location with a dark blue cursor (can maybe made thicker and a more contrasting color on the canvas), and releasing the click will move the skill to the desired location. A new "drag lock" toggle, next to the horizontal scale slider, disables click+drag behavior. Instructions have been slightly updated to describe these new interactions.

Example:
![canvas_rect_select](https://github.com/user-attachments/assets/00fef565-ed4d-4558-8b52-4f73f1fa13d2)

This PR also turns a few remaining class components into function components since I was paranoid about possible extra renders being run. Those portions of the diff can mostly be ignored; the substance of the click/drag logic happens in the the `Timeline*.tsx` family of files.

### Localization text
- Updates to the intro section instructions on clicking to toggle buffs. Added a bullet point in the instructions about click/drag, and the Timeline Editor. We should probably edit the timeline editor's help tooltip too since at this point it's a stable feature.

- Help text for the new "drag lock" setting

- Added changelog entries describing the new features.

### Stage management disclaimer

This PR is less criminal wrt state management than the invalid actions one, as the target of a drop operation is synchronized via context in the `TimelineMain` component. The `TimelineEditor` and `TimelineCanvas` will set the value of the new target index + timestamp to render a cursor when the destination of a drag operation changes.

It does introduce an additional layer of complexity in the `TimelineCanvas`. To avoid excess redraws, a second canvas is overlaid on top of the canvas that contains skill icons. This canvas is responsible for drawing the selection area box, drop target cursor, and the transparent image of the skill being dragged, and is cleared if no drag action is active. We should consider moving more "overlay-like" items, including other cursors and skill selection boxes, onto this canvas as well, but state synchronizatoin may become messy. This new canvas is updated by a `useEffect` hook monitoring the mouse position and drag state context.

### Other changes
The active timeline slot is now saved across page loads.

Fixed a bug where selecting a skill with shift-click without an active selection was a no-op (this is the one-line change in Record.ts).

An inline styling block used for the cast times table is changed to reference a `castTimesTable` class selector instead of generic `tr th td`, since for some reason after these changes, the previous inline reference to `tr th td` was affecting the editor table as well if the FPS tax table was being displayed at the same time as the editor table.

The "drag lock" and "horizontal scale" settings on the right side of the bottom tabset no longer render on smaller displays. I wanted to check a fight timeline on my phone and noticed these options were getting in the way because of absolute positioning. To be clear, I definitely do not want to first-class support mobile devices, but hiding these settings at least keeps the site from becoming actively unusable on them.

miyehn reviewed (commented):
(WIP review)

>To avoid excess redraws, a second canvas is overlaid on top of the canvas that contains skill icons.

I had the idea of separating frequently-updated canvas elements into a separate canvas for performance (not sure if my comments are still in the code somewhere), but because canvas redraws never caused any performance issues I didn't bother implementing it. So if having a second canvas is more complicated it might not be worth it for the performance gain. Fwiw it was doing a full canvas redraw every time the mouse cursor moved within the canvas, and performance was fine...? This PR is not adding a lot of elements to the canvas to make it much worse right...?

I haven't done looking at `TimelineCanvas.tsx` though, might update later today (potentially past midnight).

For now I noticed that if you end the drag within some interactable rect on the canvas (such as some skill icon, near the cursor, or some damage mark), the mouse up update is not recognized :0

zqsz-xiv commented:
> second canvas

I mostly added the second canvas overlay because I found the effect management logic for the original canvas hard to fully understand, and adding a second canvas reacting to react state changes to be easier to manage. My understanding of the canvas code is that you set flags for handling various mouse events in the canvas object's event handlers, and then would run the draw function every loop to read and reset those flags (?). I assume it was written like this to get the real-time animation mode to work properly, but these new cursors will never be drawn while in real-time mode, so I decided to just manage it separately.

There was also some functional weirdness about needing to clear the cursor in the timeline canvas when an element in the timeline editor is dragged. I probably should have added a new marker element like is done for the existing cursors so it could use existing state management logic, then just force canvas redraws on editor drag, but I didn't think about that while implementing it.

It's worth noting that while the canvas's performance is fine in the live build, but there's a noticeable amount of lag in the dev build that I believe is added by some middleware that the transpiler inserts. Eventually I'd like to make canvas reactions and animations more efficient so we can also use react's [strict mode](https://react.dev/reference/react/StrictMode) in the dev build as well, which would be even more performance overhead.

> For now I noticed that if you end the drag within some interactable rect on the canvas (such as some skill icon, near the cursor, or some damage mark), the mouse up update is not recognized :0

Should be fixed!

zqsz-xiv commented on src/Components/SkillSequencePresets.tsx:
```diff
@@ -109,75 +95,63 @@ function PresetLine(props: { line: Line }) {
+		</button>
+		<LoadJsonFromFileOrUrl
+			allowLoadFromUrl={false} // disabled for now, until when/if people have some common openers
+			defaultLoadUrl={"https://miyehn.me/ffxiv-blm-rotation/presets/lines/default.txt"}
```
I'm planning to revisit this soon-ish (#209), so when that happens I'll update this link.

miyehn reviewed (commented):
(still WIP, the rest look good but I'll probably finish looking at `TimelineCanvas.tsx` tomorrow aka today, plus maybe think about colors/styles more)

If I drag & drop edit from the timeline editor, I have the option to discard my edit in case the edit creates an invalid timeline. But drag & drop edit from the timeline itself is applied immediately? I feel like this difference may surprise users. I'm leaning toward also asking users to confirm edits before applying drag & drop from the timeline.
* The ultimate solution would be an undo stack...? But that sounds like a lot of implementation effort

Splitting canvas is fine.. but right now the second canvas also messes up the elements' z-indices, like here the selection rectangle is drawn on top of hover tooltip text but I think it's better to be the other way around:
<img width="491" height="192" alt="image" src="https://github.com/user-attachments/assets/04b7f07b-43bf-405a-b71b-739f22f0d813" />
And here 1) the drop target cursor is also on top of hover tooltip text, 2) the dotted lines are drawn outside the timeline canvas. Maybe because the second canvas is bigger than it should be?
<img width="483" height="336" alt="image" src="https://github.com/user-attachments/assets/02f4e55b-92c7-455d-9efb-1a858b54708d" />

For resolving z-index issues what immediately comes to mind is to have a 3rd canvas for whatever needs to be on top... but I'll look closer at the code changes to see if this idea makes sense

zqsz-xiv commented:
> If I drag & drop edit from the timeline editor, I have the option to discard my edit in case the edit creates an invalid timeline. But drag & drop edit from the timeline itself is applied immediately? I feel like this difference may surprise users. I'm leaning toward also asking users to confirm edits before applying drag & drop from the timeline.

The problem with this is you won't see the confirmation prompt unless you have the timeline editor tab open, so a new visitor to the site would click and drag, not realize they need to open the timeline editor to save, and then be confused when the changes aren't saved across a refresh. Drag operations aren't particularly destructive now that we're allowed to have invalid timelines, so it should be pretty easy for users to manually revert any accidental changes, though I agree an undo stack would be ideal.

> The ultimate solution would be an undo stack...? But that sounds like a lot of implementation effort
I think it actually wouldn't be that difficult, since it would just be a matter of storing skills and their positions for each timeline. Direct skill insertion/deletion in the middle of a timeline is the most requested survey feature, so when I get around to working on that I think I'll try to add an undo stack with it.

> For resolving z-index issues what immediately comes to mind is to have a 3rd canvas for whatever needs to be on top... but I'll look closer at the code changes to see if this idea makes sense

If it's too difficult to resolve I'm ok with putting all the 2nd canvas stuff I added back onto the first canvas to make sure we get layering right, and then revisit possible performance implications at a later time.

miyehn commented:
>The problem with this is you won't see the confirmation prompt unless you have the timeline editor tab open

ahh... that's a good point. Then what you have is fine. Do you want to consider also directly applying edits from the timeline editor, and only offer the option to discard if the edit is invalid? (either is fine)

zqsz-xiv commented on src/Components/TimelineCanvas.tsx:
```diff
@@ -1003,6 +1091,10 @@ function drawCursor(x: number, y1: number, y2: number, y3: number, color: string
 	g_ctx.lineTo(x, y2);
 	g_ctx.stroke();
 
+	const oldWidth = g_ctx.lineWidth;
```
I restore the original width because the dashed line for inactive timeline slots is drawn right below this, and I wanted it to remain thin.

zqsz-xiv commented:
I moved everything onto the original canvas layer, and renamed `onClick` handlers to `onMouseUp` (and made them actually fire on the canvas's `mouseUp` event).

I do think the logic is a little easier to read this way (event handler state management is separated completely from drawing logic), but these changes introduced some additional minor complications. I had to add special logic to deal with selection events in case a mouseup and mousedown were detected on the same frame, and while it's not really complicated, it's annoying to manage.

> if you care to optimize, but I doubt it matters: skillHitboxes only needs to store the positions of skill kcons, because all skill icons have the same size. And you can also only add the ones that are visible on screen

I think it's fine to keep as-is for now, since the number of skills is something like O(400) in the worst case. For detecting only those visible on screen, there's an edge case of keeping the mouse down to make a box while scrolling the canvas, which... doesn't quite work as expected right now, but I would like to eventually fix.

> Do you want to consider also directly applying edits from the timeline editor, and only offer the option to discard if the edit is invalid?

I'd like to keep status quo because the timeline editor supports individual skill deletions, which merits having a confirmation button. The end state I have in mind is to have both the timeline editor and canvas apply changes live w/o a confirm or discard prompt regardless of validity, and let people fix mistakes with an undo stack (either via ctrl-z or with a button in the editor tab).

miyehn reviewed (approved):
Proposing some colors for `dropTarget`; the current colors have really high value contrast against the background and the hue becomes hard to read

* hue closer to `lightblue`: dark theme `#55abf1` light theme `#4a9cfd`, although these may look a bit too similar to MP color especially whenthere are mp and lucid ticks present <img width="1651" height="601" alt="image" src="https://github.com/user-attachments/assets/7f168e2c-eddd-4467-b2a2-ae892d624ec8" />

* hue closer to `darkblue` (more purple-ish): dark theme `#5579f1` light theme `#4665ff` <img width="1698" height="624" alt="image" src="https://github.com/user-attachments/assets/c62ce6ed-a9d6-4799-a4b2-07659129268b" />

🚀

miyehn commented:
Oh wait, found a minor bug (?): on the timeline, if I click then shift-click to select a range of skills, the first attempt to drag and drop will fail, but subsequent drag and drops work as usual :0

zqsz-xiv commented:
I think everything works now. I'm sure people will manage to find bugs, but at least all the surface-level stuff should be good.

Behaviors that I tested, in case we need to change any logic in the future:
- Click background [deselects current selection], drag, release on background (creates selection square, selects any skills within it in the active timeline)
- Click background, release immediately (no selection is made)
With no selection:
- Click skill, release immediately (selects clicked skill)
- Shift + click skill with existing selection (expands or contracts existing selection bounds; no-op if only one skill is selected)
- Shift + click skill with no existing selection (selects clicked skill)
- Click and drag with no selection (selects clicked skill, shows target cursor, and drops it when released)
- Click and drag within existing [multi-]selection, then release on background (no-op)
- Click and drag within existing [multi-]selection, then release while hovering skill in selection (no-op)
- Click and drag with existing [multi]-selection on position near other skill, then release on background or skill (shows target cursor, then edits timeline w/o confirm prompt)

...and all these behaviors within the timeline editor table too
